modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java

```diff
@@ -221,9 +226,44 @@ private String _buildFragmentHTML(
 		return sb.toString();
 	}
 
-	private String _getAPIURL(ObjectEntry fdsEntryObjectEntry) {
-		Map<String, Object> properties = fdsEntryObjectEntry.getProperties();
+	private String _embedURLParameters(String apiUrl, HttpServletRequest httpServletRequest){
+
+		ThemeDisplay themeDisplay =
+			(ThemeDisplay)httpServletRequest.getAttribute(
+				WebKeys.THEME_DISPLAY);
+
+		Pattern pattern = Pattern.compile("\\{(.*?)\\}", Pattern.CASE_INSENSITIVE);
+		Matcher matcher = pattern.matcher(apiUrl);
+	
+		if(matcher.find()) {
+			String[] ALLOWED_PARAMETERS = {"siteId", "scopeKey", "userId"};
```

Comment:
Here we're substituting the first match only. Looks very unlikely we will have to substitute the same variable more than once, however logic should cover the case.

I feel that regexes are not very helpful in this case, because we use them for the 50% of the work. The other 50% requires checking the actual param to make the replacement.

In dxp there is a `StringUtil` class that can do that job as follows:

```
StringUtil.replace("abc/{siteid}/def/{siteid}", "{siteid}", "12345")
```

returns `abc/12345/def/12345`

So, this logic can get rid of regex management and simplified as a sequence of replacements for the known tokens. Moreover there is a [variant of that method](https://github.com/liferay/liferay-portal/blob/451c0f4cc61be5c9476b79d93870b9637320dcbc/portal-kernel/src/com/liferay/portal/kernel/util/StringUtil.java#L2852) that takes 2 arrays, one with the tokens and the other with the values, making it possible to do all the substitutions in a shot.

This will work fine as long as the values don't include any token, which is unlikely by design.

Reply:
The String.replace method changes all occurrences in the string (tested with crazy urls), unless I missed something here.

I would agree to get rid of the entire regex matcher logic and go for a "Tell, don't ask" pattern. In other words, always try to make a replacement, no matter if there is something to be replaced or not, in order to reduce the cognitive load of the code.